[QA] Add unit tests and docs for `DatePicker` (#319)
mckinsey/vizro: vizro-core/docs/pages/user-guides/selectors.md

the [`RadioItem`][vizro.models.RadioItems] serves as a single-option selector by default. However, the
[`Dropdown`][vizro.models.Dropdown] can function as both a multi-option or single-option selector.

> For more details, refer to the documentation of the underlying Dash components:

Reviewer: Sorry to drag this out, but I have decided that I do not like the presentation here. What bugs me, and I have actively tripped over this, is that these are the docs about our selectors, but the most prominent links are to the underlying components.

What do you think of slightly restructuring this such that the bullets are smth a long those line:

```
For more information, refer to the API reference of the selector, or the documentation of its underlying Dash component:
- [`Checklist`][vizro.models.Checklist] based on [`dcc.Dropdown`](https://dash.plotly.com/dash-core-components/dropdown)
- ...
```

!!! note

    When configuring the [`Slider`][vizro.models.Slider] and the [`RangeSlider`][vizro.models.RangeSlider] with float values, and using `step` with an integer value, you may notice
    unexpected behavior, such as the drag value being outside its indicated marks.
    To our knowledge, this is a current bug in the underlying [`dcc.Slider`](https://dash.plotly.com/dash-core-components/slider) and
    [`dcc.RangeSlider`](https://dash.plotly.com/dash-core-components/rangeslider) component, which you can circumvent by adapting the `step` size accordingly.

To utilize `numerical` [`Filter`][vizro.models.Filter] selectors, the filtered column must be of `numeric` dtype,
indicating that [pandas.api.types.is_numeric_dtype()](https://pandas.pydata.org/docs/reference/api/pandas.api.types.is_numeric_dtype.html) must return `True` for the filtered column.

## Temporal Selectors

For more details, refer to the documentation of the underlying Dash components:

- [`dmc.DateRangePicker`](https://www.dash-mantine-components.com/components/datepicker#daterangepicker)
- [`dmc.DatePicker`](https://www.dash-mantine-components.com/components/datepicker)

!!! note

>     [`DatePicker`][vizro.models.DatePicker] component is based on two underlying components `dmc.DatePicker` and `dmc.DateRangePicker`.
>     When [`DatePicker`][vizro.models.DatePicker] is configured with `range=True` (the default), the underlying component is `dmc.DateRangePicker`; when `range=False` the component is `dmc.DatePicker`.
>     When configuring the [`DatePicker`][vizro.models.DatePicker] make sure to provide your dates for `min`, `max` and `value` arguments in `"yyyy-mm-dd"` format or as datetime type.

Reviewer: ```suggestion
    The [`DatePicker`][vizro.models.DatePicker] component is based on two underlying components -  `dmc.DatePicker` and `dmc.DateRangePicker`.
    When the [`DatePicker`][vizro.models.DatePicker] is configured with `range=True` (the default), the underlying component is `dmc.DateRangePicker`. When `range=False` the underlying component is `dmc.DatePicker`.
    When configuring the [`DatePicker`][vizro.models.DatePicker] make sure to provide your dates for `min`, `max` and `value` arguments in `"yyyy-mm-dd"` format or as `datetime` type.
```
Can we add what `datatime` type is with a link. Like convertible by xy function or so

Reviewer 2: Can you take a quick look at the docs and see if everything is fine now and if I still need to add a datetime type link?

> To utilize `temporal` [`Filter`][vizro.models.Filter] selectors, the filtered column must be of `datetime` dtype,

Reviewer: ```suggestion
To utilize temporal [`Filter`][vizro.models.Filter] selectors, the filtered column must be of `datetime` format,
```

indicating that [pandas.api.types.is_datetime64_any_dtype()](https://pandas.pydata.org/docs/reference/api/pandas.api.types.is_datetime64_any_dtype.html) must return `True` for the filtered column.

> !!! note

Reviewer: ```suggestion
!!! tip
```

>     Dataframe column type can be changed to datetime using:

Reviewer: ```suggestion
    `pd.DataFrame` column types can be changed to `datetime` using:
```
but please check this again, currently it renders ugly

Reviewer: maybe even skip this, or make this a one liner with the first option only for simplification

     - [pandas.to_datetime()](https://pandas.pydata.org/docs/reference/api/pandas.to_datetime.html) or
     - [pandas.DataFrame.astype()](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.astype.html) with `datetime64` as the argument.

> 

Reviewer: ```suggestion

## Example of default Filter selectors
```

> Below is an example where all selectors are automatically set by default based on the type of the filtered column.

Reviewer 3: ```suggestion
### Default selectors

If you don't specify a selector, a default selector is applied based on the data type of the provided data column. 

Default selectors for:
- categorical data: vm.Dropdown
- numerical data: vm.RangeSlider
- temporal data: vm.DatePicker(range=True)

...
```

Reviewer 3: Below that section you could add some information of how we identify categorical/numerical/temporal data, so basically what you have above with the pandas.api.types etc.

> !!! example "Default Filter selectors"

Reviewer: Great idea of an example! I would use a well known DF though, so the yaml version is easier repeatable?

Reviewer 2: There is no well known DF [here](https://plotly.com/python-api-reference/generated/plotly.express.data.html) that contains fine temporal, categorical and numerical columns together.

The closest we have is `px.data.gapminder(datetimes=True)` which we avoid using (https://github.com/mckinsey/vizro/pull/319#discussion_r1525060748) in the examples since datetimes are around years with `Jan-01`. So to represent DatePicker selector we use `px.data.stocks` DF in the examples (e.g. features examples).

Reviewer: And in stocks, we are missing the categorical column right?

Reviewer 3: You can transform the stocks data set by changing it from wide to long, so the columns 'GOOG', 'AAPL', 'AMZN', 'FB', 'NFLX', 'MSFT' become categories in a column we call "Stocks" and then you have a numeric column with the values :)

Then you have a data set with date, categorical and numeric values.

## Numerical selectors

For more details, refer to the documentation of the underlying Dash components:

- [`dcc.Slider`](https://dash.plotly.com/dash-core-components/slider])
- [`dcc.RangeSlider`](https://dash.plotly.com/dash-core-components/rangeslider])

!!! note

    When configuring the [`Slider`][vizro.models.Slider] and the [`RangeSlider`][vizro.models.RangeSlider] with float values, and using `step` with an integer value, you may notice
    unexpected behavior, such as the drag value being outside its indicated marks.
    To our knowledge, this is a current bug in the underlying [`dcc.Slider`](https://dash.plotly.com/dash-core-components/slider) and
    [`dcc.RangeSlider`](https://dash.plotly.com/dash-core-components/rangeslider) component, which you can circumvent by adapting the `step` size accordingly.

## Temporal Selectors

For more details, refer to the documentation of the underlying Dash components:

- [`dmc.DateRangePicker`](https://www.dash-mantine-components.com/components/datepicker#daterangepicker)
- [`dmc.DatePicker`](https://www.dash-mantine-components.com/components/datepicker)

!!! note

>     The [`DatePicker`][vizro.models.DatePicker] component is based on two underlying components -  `dmc.DatePicker` and `dmc.DateRangePicker`.

Reviewer: This still renders weird for me:
![Uploading image.png…]()

Reviewer: If we follow this guidance (https://github.com/mckinsey/vizro/pull/319/files#r1528241690), then we could also shorting this note significantly. I would then just add this explanation below the bullet point about this selector.

The note should then contain the only in my eyes important thing for the reader to note, which is the need to provide the correct date formats.

Given that this section is before the other, I wonder still how to make it clear that `datetime` actually is

Reviewer: Yes, that looks ok! Although my comment about shorting this still holds :)

    When the [`DatePicker`][vizro.models.DatePicker] is configured with `range=True` (the default), the underlying component is `dmc.DateRangePicker`. When `range=False` the underlying component is `dmc.DatePicker`.
    When configuring the [`DatePicker`][vizro.models.DatePicker] make sure to provide your dates for `min`, `max` and `value` arguments in `"yyyy-mm-dd"` format or as `datetime` type.

## Default selectors

> If you don't specify a selector, a default selector is applied based on the data type of the provided data column.
> 

Reviewer: ```suggestion
If you don't specify a selector, a default selector is applied based on the data type of the provided column.

```

Default selectors for:

>  - categorical data: vm.Dropdown
>  - numerical data: vm.RangeSlider
>  - temporal data: vm.DatePicker(range=True)

Reviewer: ```suggestion
 - categorical data: [`Dropdown`][vizro.models.Dropdown]
 - numerical data: [`RangeSlider`][vizro.models.RangeSlider]
 - temporal data: [`DatePicker(range=True)`][vizro.models.DatePicker]```

Categorical selectors can be used independently of the data type of the column being filtered.

> To utilize numerical [`Filter`][vizro.models.Filter] selectors, the filtered column must be of `numeric` format,

Reviewer 4: ```suggestion
To use numerical [`Filter`][vizro.models.Filter] selectors, the filtered column must be of `numeric` format,
```

indicating that [pandas.api.types.is_numeric_dtype()](https://pandas.pydata.org/docs/reference/api/pandas.api.types.is_numeric_dtype.html) must return `True` for the filtered column.

> To utilize temporal [`Filter`][vizro.models.Filter] selectors, the filtered column must be of `datetime` format,

Reviewer 4: ```suggestion
To use temporal [`Filter`][vizro.models.Filter] selectors, the filtered column must be of `datetime` format,
```

## Numerical selectors

For more details, refer to the documentation of the underlying dash components:

- [dcc.Slider](https://dash.plotly.com/dash-core-components/slider])
- [dcc.RangeSlider](https://dash.plotly.com/dash-core-components/rangeslider])

!!! note

    When configuring the [`Slider`][vizro.models.Slider] and the [`RangeSlider`][vizro.models.RangeSlider] with float values, and using `step` with an integer value, you may notice
    unexpected behavior, such as the drag value being outside its indicated marks.
    To our knowledge, this is a current bug in the underlying [`dcc.Slider`](https://dash.plotly.com/dash-core-components/slider) and
    [`dcc.RangeSlider`](https://dash.plotly.com/dash-core-components/rangeslider) component, which you can circumvent by adapting the `step` size accordingly.

## Temporal Selectors

For more details, refer to the documentation of the underlying dash components:

> - [dmc.DateRangePicker](https://www.dash-mantine-components.com/components/datepicker#daterangepicker)
> - [dmc.DatePicker](https://www.dash-mantine-components.com/components/datepicker)

Reviewer 5: ```suggestion
- [`dmc.DateRangePicker`](https://www.dash-mantine-components.com/components/datepicker#daterangepicker)
- [`dmc.DatePicker`](https://www.dash-mantine-components.com/components/datepicker)
```

!!! note

>     [`DatePicker`][vizro.models.DatePicker] component is based on 2 underlying components dmc.DatePicker and dmc.DateRangePicker.
>     When [`DatePicker`][vizro.models.DatePicker] is configured with `range=False` argument, underlying component is dmc.DatePicker. Default value of `range` is `True`,
>     meaning [`DatePicker`][vizro.models.DatePicker] by default uses dmc.DateRangePicker.
>     When configuring the [`DatePicker`][vizro.models.DatePicker] make sure to provide your dates for `min`, `max` and `value` arguments in `"yyyy-mm-dd"` format or as datetime type.

Reviewer 5: ```suggestion
    [`DatePicker`][vizro.models.DatePicker] component is based on two underlying components `dmc.DatePicker` and `dmc.DateRangePicker`.
    When [`DatePicker`][vizro.models.DatePicker] is configured with `range=True` (the default), the underlying component is `dmc.DateRangePicker`; when `range=False` the component is `dmc.DatePicker`.
    When configuring the [`DatePicker`][vizro.models.DatePicker] make sure to provide your dates for `min`, `max` and `value` arguments in `"yyyy-mm-dd"` format or as datetime type.
```

It almost always makes sense to describe the default behaviour first.
